packages/neos-ui/src/Containers/Modals/AddNode/index.js

```diff
@@ -32,91 +67,235 @@ export default class AddNodeModal extends Component {
         referenceNodeParent: NeosPropTypes.node,
         referenceNodeGrandParent: NeosPropTypes.node,
         groupedAllowedNodeTypes: PropTypes.array,
-        mode: PropTypes.string.isRequired,
+        allowedNodeTypesByModeGeneratorFn: PropTypes.func.isRequired,
         nodeTypesRegistry: PropTypes.object.isRequired,
 
-        close: PropTypes.func.isRequired
+        handleClose: PropTypes.func.isRequired,
+        persistChange: PropTypes.func.isRequired
     };
 
-    shouldComponentUpdate(newProps, newState) {
-        return shallowCompare(this, newProps, newState) && newProps.referenceNode !== this.props.referenceNode;
+    shouldComponentUpdate(...args) {
+        return shallowCompare(this, ...args);
+    }
+
+    componentWillReceiveProps(nextProps) {
+        if (this.props.referenceNode !== nextProps.referenceNode) {
+            this.setState({step: 1});
+        }
+    }
+
+    constructor(...props) {
+        super(...props);
+
+        this.state = {
+            mode: 'insert',
+            step: 1,
+            selectedNodeType: null,
+            propertiesOfToBeCreatedNode: {}
+        };
+
+        this.handleSelectNodeType = this.handleSelectNodeType.bind(this);
+        this.handleModeChange = this.handleModeChange.bind(this);
+        this.handleDialogEditorValueChange = this.handleDialogEditorValueChange.bind(this);
+        this.handleSave = this.handleSave.bind(this);
     }
 
     render() {
+        if (!this.props.referenceNode) {
+            return null;
+        }
+
+        if (this.state.step === 1) {
+            return this.renderStep1();
```

Comment:
I'd prefer more speaking names for the steps. Also constants instead of literals (`this.state.step === STEP_CHOOSE_MODE` for instance). So we get a clear terminology of what this workflow is about.